firedrake/ufl_expr.py

```diff
@@ -138,15 +138,24 @@ def derivative(form, u, du=None, coefficient_derivatives=None):
 
     See also :func:`ufl.derivative`.
     """
-    if set(extract_coefficients(form)) & set(u.split()):
+    # TODO: What about Constant?
+    if len(u.split()) > 1 and set(extract_coefficients(form)) & set(u.split()):
```

Comment:
This check is ugly.  If `f` is not a mixed function, then `f.split()` is a singleton tuple, which contains a new function that shares data with `f`.  Whereas if `f` is a `Constant`, then `f.split()` is a singleton tuple which is `f`.